test/lit/passes/code-folding-eh-old.wast

```diff
@@ -116,7 +159,145 @@
+  ;; CHECK-NEXT:    (drop
+  ;; CHECK-NEXT:     (i32.const 1)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (return
+  ;; CHECK-NEXT:     (call $foo-i32)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (i32.const 0)
+  ;; CHECK-NEXT: )
+  (func $try-call-optimize-terminating-tails-call-return (result i32)
+    (try
+      (do
+        (drop (i32.const 1))
+        (drop (i32.const 1))
+        ;; Cannot be folded out of the try because it might throw.
+        (return (call $foo-i32))
+      )
+      (catch_all
+        (drop (i32.const 1))
+        (drop (i32.const 1))
+        (return (call $foo-i32))
+      )
+    )
+    (i32.const 0)
+  )
+
+  ;; CHECK:      (func $try-call-optimize-terminating-tails-return-call (type $0) (result i32)
+  ;; CHECK-NEXT:  (block $folding-inner0
+  ;; CHECK-NEXT:   (return
+  ;; CHECK-NEXT:    (block (result i32)
+  ;; CHECK-NEXT:     (try
+  ;; CHECK-NEXT:      (do
+  ;; CHECK-NEXT:       (br $folding-inner0)
+  ;; CHECK-NEXT:      )
+  ;; CHECK-NEXT:      (catch_all
+  ;; CHECK-NEXT:       (br $folding-inner0)
+  ;; CHECK-NEXT:      )
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:     (i32.const 0)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.const 1)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.const 1)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.const 1)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (return_call $foo-i32)
+  ;; CHECK-NEXT: )
+  (func $try-call-optimize-terminating-tails-return-call (result i32)
+    (try
+      (do
+        (drop (i32.const 1))
+        (drop (i32.const 1))
+        (drop (i32.const 1))
+        (return_call $foo-i32)
+      )
+      (catch_all
+        (drop (i32.const 1))
+        (drop (i32.const 1))
+        (drop (i32.const 1))
+        (return_call $foo-i32)
+      )
+    )
+    (i32.const 0)
+  )
+
+  ;; CHECK:      (func $try-call-optimize-expression-tails-success (type $1)
+  ;; CHECK-NEXT:  (block $x
+  ;; CHECK-NEXT:   (try
+  ;; CHECK-NEXT:    (do
+  ;; CHECK-NEXT:     (br $x)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (catch_all
+  ;; CHECK-NEXT:     (br $x)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (unreachable)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.const 1)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.const 1)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.const 1)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $try-call-optimize-expression-tails-success
+    (block $x
+      (try
+        (do
+          (drop (i32.const 1))
+          (drop (i32.const 1))
+          (drop (i32.const 1))
+          (br $x)
+        )
+        (catch_all
+          (drop (i32.const 1))
+          (drop (i32.const 1))
+          (drop (i32.const 1))
+          (br $x)
+        )
+      )
+      (unreachable)
+    )
+  )
```

Comment:
did we not have a test for this before?

Reply:
No, not that I was able to find for either of these tests. Except for the throws effect analysis, the implementation does not seem to special case try-catch at all, so it's reasonable that we wouldn't need tests for the success cases. But when I was trying to debug and figure out how to get this change working, it was useful to prove to myself locally that similar cases did already work correctly.